### PR TITLE
Parameter preconditions and typereader overriding

### PR DESCRIPTION
**Pull request by FiniteReality**
Using the new command builder system, since I'm not sure whether Joe4evr is going to update his system.

Parameter preconditions had to be done as part of the execute phase as they require the value passed to the command after parsing - I wasn't sure about the best way to include it as a separate phase as it would have meant changing how the parameter list was built and passed around heavily.

### CODE REVIEW

**Comment by FiniteReality** [comment]
So, I'm a derp and didn't test properly; the latest commit fixes the mistakes I made.

File: src/Discord.Net.Commands/Builders/ModuleClassBuilder.cs [hunk 1]
```diff
@@ -193,23 +199,47 @@ private static void BuildParameter(ParameterBuilder builder, System.Reflection.P
...
             }
 
-            var reader = service.GetTypeReader(paramType);
-            if (reader == null)
+            if (builder.TypeReader == null)
             {
-                var paramTypeInfo = paramType.GetTypeInfo();
-                if (paramTypeInfo.IsEnum)
+                var readers = service.GetTypeReaders(paramType);
+                var reader = readers?.FirstOrDefault();
```
Comment (RogueException) [inline src/Discord.Net.Commands/Builders/ModuleClassBuilder.cs]: Is it safe to always assume the first reader is the default? Maybe we should have a separate field for the default/global reader?

Comment (RogueException) [inline src/Discord.Net.Commands/Builders/ModuleClassBuilder.cs]: It seems you're using ConcurrentBags which, by definition, do not guarantee any ordering.

File: src/Discord.Net.Commands/Info/CommandInfo.cs [hunk 2]
```diff
@@ -125,9 +125,26 @@ public Task<ExecuteResult> Execute(CommandContext context, ParseResult parseResu
             if (map == null)
                 map = DependencyMap.Empty;
 
+            object[] args = null;
+
+            try
```
Comment (RogueException) [inline src/Discord.Net.Commands/Info/CommandInfo.cs]: It's safe to leave this try-catch around the whole op, instead of splitting it into two sections.

**Review by Auralytical — CHANGES_REQUESTED** [review]

**Review by Auralytical — APPROVED** [review]

**Comment by RogueException** [comment]
If this has been retested, PR is approved once conflicts are resolved.

**Comment by FiniteReality** [comment]
@RogueException Any news on merging this? As far as I'm aware this is ready

**Pull request merged**